dds/DCPS/DataReaderImpl.cpp

```diff
@@ -304,6 +305,21 @@ DataReaderImpl::add_association(const RepoId& yourId,
     }
   }
 
+  //get message block from octet seq then deser to a type info
+  ACE_Data_Block db(writer.serializedTypeInfo.length(), ACE_Message_Block::MB_DATA,
+                    reinterpret_cast<const char*>(writer.serializedTypeInfo.get_buffer()),
+                    0 /*alloc*/, 0 /*lock*/, ACE_Message_Block::DONT_DELETE, 0 /*db_alloc*/);
+  ACE_Message_Block data_mb(&db, ACE_Message_Block::DONT_DELETE, 0 /*mb_alloc*/);
+  data_mb.wr_ptr(data_mb.space());
+  DCPS::Serializer serializer(&data_mb, XTypes::get_typeobject_encoding());
+  XTypes::TypeInformation type_info;
+  if (!(serializer >> type_info)) {
+    ACE_ERROR((LM_ERROR, ACE_TEXT("(%P|%t) add_association ")
+              ACE_TEXT("deserialization type information failed.\n")));
+  }
```

Comment:
This is the part I mentioned before about integrating both RTPS Discovery and DCPSInfoRepo Discovery.

The main issue is that currently the "add_association" call from Discovery to Reader/Writer doesn't pass the TypeInformation along with it.  There are a few ways this could be fixed but I think they basically boil down to either adding an additional parameter to `DataReaderCallbacks::add_association` or changing the `WriterAssocation` parameter from being an IDL struct to a C++ struct that wraps that IDL struct and adds TypeInformation.  (Same for DataWriterCallbacks and ReaderAssociation).

Once that is resolved, the code tagged in this review (converting OctetSeq to TypeInformation) can either be moved to a single location that's naturally in the control flow for both RTPS Discovery and DCPSInfoRepo Discovery, or turned in to a utility function that's called from both.